contrib/rust/src/python/pants/contrib/rust/utils/basic_invocation_conversion.py

```diff
@@ -0,0 +1,56 @@
+# coding=utf-8
+# Copyright 2019 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import absolute_import, division, print_function, unicode_literals
+
+import copy
+
+from pants.contrib.rust.tasks.cargo_task import CargoTask
+from pants.contrib.rust.utils.basic_invocation.args_rules import args_rules
+from pants.contrib.rust.utils.basic_invocation.env_rules import env_rules
+from pants.contrib.rust.utils.basic_invocation.links_rule import links_rule
+from pants.contrib.rust.utils.basic_invocation.outputs_rule import outputs_rule
+from pants.contrib.rust.utils.basic_invocation_conversion_utils import (reduce_invocation,
+                                                                        sanitize_crate_name)
+from pants.contrib.rust.utils.collector import (collect_information, get_default_information,
+                                                get_test_target_information)
+
+
+def get_default_conversion_rules():
```

Comment:
I don't think I understand the pattern behind the `rules` that are exposed by `basic_invocation` and `custom_invocation`.

Are they intended to have the same shape (similar to two implementations of a trait/interface)? If so, should something enforce/signal that (ie: should they be in an interface)? If not, some comments to explain their composition would be helpful.

And did they start out as `@rules` as in https://github.com/pantsbuild/pants/blob/master/src/python/pants/engine/README.md ? If so, would be interested to hear if you ran into trouble using that model, as we're still iterating on it, and feedback would be welcome.